File: app/src/test/java/com/coinninja/coinkeeper/db/DatabaseSecretProviderTest.java
package com.coinninja.coinkeeper.db;

import android.os.Build;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.coinninja.coinkeeper.util.Hasher;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
public class DatabaseSecretProviderTest {

    private String appSecret = "--app-secret--";
    private String defaultSecret = "--default-secret--";

    @Mock
    private Hasher hasher;

    private DatabaseSecretProvider databaseSecretProvider;

    @After
    public void tearDown() {
        appSecret = null;
        defaultSecret = null;
        hasher = null;
        databaseSecretProvider = null;
    }

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        databaseSecretProvider = new DatabaseSecretProvider(hasher, appSecret, defaultSecret);
    }

    @Test
    @Config(sdk = Build.VERSION_CODES.O)
    public void hashes_app_secret() {
        String hashedSecret = "--hashed-secret--";
        when(hasher.hash(appSecret)).thenReturn(hashedSecret);

        assertThat(databaseSecretProvider.getSecret(), equalTo(hashedSecret));
    }

    @Test
    @Config(sdk = Build.VERSION_CODES.N)
    public void uses_default_on_lower_apis() {
        assertThat(databaseSecretProvider.getSecret(), equalTo(defaultSecret));
    }

    @Test
    public void provides_default() {
        String hashedSecret = "--hashed-secret--";
        when(hasher.hash(appSecret)).thenReturn(hashedSecret);

        assertThat(databaseSecretProvider.getDefault(), equalTo(defaultSecret));
    }


}

File: app/src/test/java/com/coinninja/coinkeeper/db/DatabaseSecretProviderTest.kt
package com.coinninja.coinkeeper.db

import android.os.Build.VERSION_CODES
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config
import java.io.Serializable

@RunWith(AndroidJUnit4::class)
class DatabaseSecretProviderTest {
    companion object {
        private val appSecret: CharArray = "--app-secret--".toCharArray()
        private val defaultSecret: CharArray = "--default-secret--".toCharArray()

    }

    private fun createSecretProvider(): DatabaseSecretProvider = DatabaseSecretProvider(mock(), appSecret, defaultSecret)

    @Test
    @Config(sdk = [VERSION_CODES.O])
    fun hashes_app_secret() {
        val hashedSecret = "--hashed-secret--".toCharArray()
        val provider = createSecretProvider()
        whenever(provider.hasher.hash(appSecret)).thenReturn(hashedSecret)
        assertThat(provider.secret, equalTo<Serializable>(hashedSecret))
    }

    @Test
    @Config(sdk = [VERSION_CODES.N])
    fun uses_default_on_lower_apis() {
        val provider = createSecretProvider()
        assertThat(provider.secret, equalTo<Serializable?>(defaultSecret))
    }

    @Test
    fun provides_default() {
        val provider = createSecretProvider()
        val hashedSecret = "--hashed-secret--".toCharArray()
        whenever(provider.hasher.hash(appSecret)).thenReturn(hashedSecret)
        assertThat(provider.default, equalTo(defaultSecret))
    }
}